vllm/ex/fusion.py

```diff
@@ -0,0 +1,284 @@
+
+    # assumption, graph.nodes are in topo order
+    mod.graph.lint()
+
+    logger.debug("start fusion")
+
+    # create partition groups
+    # run in reverse order so predecesors of non-unary ops will appear
+    # in the same partition.
+    for n in reversed(mod.graph.nodes):
+        logger.debug(f"CONSIDER {n}")
+
+        if not is_call(n):
+            logger.debug(f"  REJECT {n} not call")
+            node_map[n] = 0
+            continue
+
+        # TODO: handle get_attr ops
+        # should probably be lifted/put in partition 0 but not prevent fusion
+
+        pred = is_fusable_pair if not fuse_with_compute else is_compute_fusable_pair
+
+        fusable = [pred(s, n) for s in fg.predecessors(n) if s.op != 'placeholder']
+        if not all(fusable):
+            if not n in node_map:
+                logger.debug(f"  REJECT {n} no fusable preds and not in map: {fusable}, {fg.predecessors(n)}")
+                node_map[n] = 0
+            continue
+
+        # don't support anything with kwargs for now
+        if not supported_kwargs(n):
+            logger.debug(f"  REJECT {n} unsupported kwargs")
+            node_map[n] = 0
+            continue
+
+        if n not in node_map:
+            partition = partition + 1
+            node_map[n] = partition
+
+        for s in fg.predecessors(n):
+            if s.op != 'placeholder':
+                node_map[s] = node_map[n]
+
+    logger.debug(f"node_map = {node_map}")
+
+    def same_partition(nodes: Set[torch.fx.Node]) -> bool:
+        if len(nodes) > 0:
+            part = node_map[next(iter(nodes))]
+            return all([node_map[n] == part for n in nodes])
+        return False
+
+    def only_pointwise(partition: int) -> bool:
+        nodes = [n for n, p in node_map.items() if p == partition]
+        return all([is_fusable(n) and not is_compute(n) for n in nodes])
+
+
+    if fuse_with_compute:
+        for n in mod.graph.nodes:
+            if not is_call(n):
+                continue
+
+            if fuse_inputs:
+                nodes = fg.predecessors(n)
+            else:
+                nodes = fg.successors(n)
+
+            if not is_compute(n):
+                continue
+
+            if not same_partition(nodes):
+                #logger.debug(f"REJECT {n} not all neighbors in same partition {nodes}")
+                continue
+
+            fuse_part = next(iter(nodes))
+
+            if only_pointwise(fuse_part):
+                node_map[n] = node_map[fuse_part]
+
+    logger.debug(f"final node_map = {node_map}")
+
+    assert(all([n in node_map for n in mod.graph.nodes]))
+
+    logger.debug(f"pre-fusion split mod:\n{graph_print_tabular(mod.graph, 'part', map_node)}")
+
+    subgraphs = dict()
+    for n, p in node_map.items():
+        if p > 0:
+            if not p in subgraphs:
+                subgraphs[p] = []
+            subgraphs[p].append(n)
+
+    for p, nodes in subgraphs.items():
+        sub = SubGraph(mod, subgraphs[p])
```

Comment:
This may just be my unfamiliarity with fx graphs, but I think it would be good to comment that changes made to the SubGraph will also be made to the `mod`